merkletree/merkletree.go

```diff
@@ -0,0 +1,288 @@
+// Asking for the root when no data has been added will return nil. The tree is
+// left unaltered.
+func (t *Tree) Root() (root []byte) {
+	// If the Tree is empty, return nil.
+	if t.head == nil {
+		return nil
+	}
+
+	// The root is formed by hashing together subTrees in order from least in
+	// height to greatest in height. To preserve ordering, the larger subTree
+	// needs to be first in the combination.
+	current := t.head
+	root = current.sum
+	for current.next != nil {
+		root = join(t.hash, current.next.sum, root)
+		current = current.next
+	}
+	return root
+}
+
+// Prove returns a proof that the data at index 'proveIndex' is an element in
+// the current Tree. The proof will be invalid if any more elements are added
+// to the tree after calling Prove. The tree is left unaltered.
+func (t *Tree) Prove() (h hash.Hash, merkleRoot []byte, proveSet [][]byte, proveIndex int, numLeaves int) {
+	// Return nil if the Tree is empty, or if the proveIndex hasn't yet been
+	// reached.
+	if t.head == nil || len(t.proveSet) == 0 {
+		return h, t.Root(), nil, t.proveIndex, t.currentIndex
+	}
+	proveSet = t.proveSet
+
+	// The hashes have already been provided for the largest complete subTree
+	// that contains 't.ProveIndex'. If 't.CurrentIndex' is a power of two, we
+	// are already finshed. Otherwise, two sets of hashes remain which need to
+	// be added to the proof. The first is the hashes of the smaller subTrees.
+	// All of the smaller subTrees need to be combined, and then that hash
+	// needs to be saved. The second is the larger subTrees. The root of each
+	// of the larger subTrees needs to be saved. The subTree with the prove
+	// index will have a height equal to the current length of the prove set.
+
+	// Iterate through all of the smaller subTrees and combine them.
+	current := t.head
+	sum := current.sum
+	for current.next != nil && current.next.height < len(proveSet) {
+		// Combine this subTree with the next subTree.
+		sum = join(t.hash, current.next.sum, sum)
+		current = current.next
+	}
+
+	// If the current subTree is the last subTree before the subTree containing
+	// the prove index, add the root of the subTree to the prove set.
+	if current.next != nil && current.next.height == len(proveSet) {
+		proveSet = append(proveSet, sum)
+		current = current.next
+	}
+
+	// The subTree containing the prove index needs to be skipped.
+	current = current.next
+
+	// Now add the roots of all subTrees that are larger than the subTree
+	// containing the proof index.
+	for current != nil {
+		proveSet = append(proveSet, current.sum)
+		current = current.next
+	}
+	return t.hash, t.Root(), proveSet, t.proveIndex, t.currentIndex
+}
+
+// VerifyProof takes a merkle, a proveSet, and a proveIndex and returns true if
+// the first element of the prove set is a leaf of data in the merkle root.
+func VerifyProof(h hash.Hash, merkleRoot []byte, proveSet [][]byte, proveIndex int, numLeaves int) bool {
```

Comment:
lol I did it the other way. s/proof/prove

heh. I'll let you do whatever naming changes and other finishing touches. I think the test suite is nearly comprehensive, so as long as that passes you shouldn't unintentionally introduce any errors.